syntax/pretty_print.cpp

```diff
@@ -3,365 +3,374 @@
-      print_extern(dynamic_cast<ast::Extern*>(node));
+      print_extern(stream, dynamic_cast<const ast::Extern&>(node));
       break;
     case ast::FuncNode:
-      print_func(dynamic_cast<ast::Func*>(node));
+      print_func(stream, dynamic_cast<const ast::Func&>(node));
       break;
     case ast::VariableNode:
-      print_variable(dynamic_cast<ast::Variable*>(node));
+      print_variable(stream, dynamic_cast<const ast::Variable&>(node));
       break;
     case ast::ConstantNode:
-      print_constant(dynamic_cast<ast::Constant*>(node));
+      print_constant(stream, dynamic_cast<const ast::Constant&>(node));
       break;
     case ast::CastNode:
-      print_cast(dynamic_cast<ast::Cast*>(node));
+      print_cast(stream, dynamic_cast<const ast::Cast&>(node));
       break;
     case ast::UnOpNode:
-      print_un_op(dynamic_cast<ast::UnOp*>(node));
+      print_un_op(stream, dynamic_cast<const ast::UnOp&>(node));
       break;
     case ast::BinOpNode:
-      print_bin_op(dynamic_cast<ast::BinOp*>(node));
+      print_bin_op(stream, dynamic_cast<const ast::BinOp&>(node));
       break;
     case ast::CallNode:
-      print_call(dynamic_cast<ast::Call*>(node));
+      print_call(stream, dynamic_cast<const ast::Call&>(node));
       break;
     case ast::ReturnNode:
-      print_return(dynamic_cast<ast::Return*>(node));
+      print_return(stream, dynamic_cast<const ast::Return&>(node));
       break;
     case ast::BranchNode:
-      print_branch(dynamic_cast<ast::Branch*>(node));
+      print_branch(stream, dynamic_cast<const ast::Branch&>(node));
       break;
     case ast::AssignmentNode:
-      print_assignment(dynamic_cast<ast::Assignment*>(node));
+      print_assignment(stream, dynamic_cast<const ast::Assignment&>(node));
       break;
     case ast::WhileNode:
-      print_while(dynamic_cast<ast::While*>(node));
+      print_while(stream, dynamic_cast<const ast::While&>(node));
       break;
     case ast::VoidContextNode:
-      print_void_context(dynamic_cast<ast::VoidContext*>(node));
+      print_void_context(stream, dynamic_cast<const ast::VoidContext&>(node));
       break;
     default:
-      std::cout << "[Unknown Node] ??";
-      return EXIT_FAILURE;
+      stream << "[Unknown Node] ??";
     }
-    return EXIT_SUCCESS;
+    return stream;
   }
 
-  int program(ast::Program* program) {
-    assert(program != nullptr);
-    for (auto it = program->body.begin(); it != program->body.end(); ++it) {
-      node(it->get());
+  std::ostream& program(std::ostream& stream, const ast::Program& program) {
+    for (auto it = program.body.begin(); it != program.body.end(); ++it) {
+      node(stream, *(it->get()));
     }
-    return EXIT_SUCCESS;
+    return stream;
   }
 };
 
-int pretty_print::node(ast::abstract::Node* node) {
-  return PrettyPrinter().node(node);
+std::ostream& pretty_print::node(std::ostream& stream,
+        const ast::abstract::Node& node) {
+  return PrettyPrinter().node(stream, node);
```

Comment:
`return PrettyPrinter(stream).node(node);` -> this way you don't have to put extra argument to every function